src/main/java/io/openraven/magpie/plugins/gcp/discovery/services/ServiceDirectoryDiscovery.java

```diff
@@ -0,0 +1,157 @@
+    final String RESOURCE_TYPE = "GCP::ServiceDirectory::Namespace";
+
+    try (RegistrationServiceClient registrationServiceClient = RegistrationServiceClient.create()) {
+      AVAILABLE_LOCATIONS.forEach(location -> {
+        String parent = LocationName.of(projectId, location).toString();
+
+        registrationServiceClient.listNamespaces(parent).iterateAll().forEach(namespace -> {
+          var data = new MagpieResource.MagpieResourceBuilder(mapper, namespace.getName())
+            .withProjectId(projectId)
+            .withResourceType(RESOURCE_TYPE)
+            .withConfiguration(GCPUtils.asJsonNode(namespace))
+            .build();
+
+          emitter.emit(VersionedMagpieEnvelopeProvider.create(session, List.of(fullService() + ":namespace"), data.toJsonNode()));
+
+          discoverServices(mapper, projectId, session, emitter, registrationServiceClient, location,
+            getEntityNameFromPath(namespace.getName()));
+        });
+      });
+    } catch (IOException e) {
+      DiscoveryExceptions.onDiscoveryException(RESOURCE_TYPE, e);
+    }
+  }
+
+  private void discoverServices(ObjectMapper mapper,
+                                String projectId,
+                                Session session,
+                                Emitter emitter,
+                                RegistrationServiceClient registrationServiceClient,
+                                String location,
+                                String namespaceName) {
+    final String RESOURCE_TYPE = "GCP::ServiceDirectory::Service";
+
+    registrationServiceClient.listServices(NamespaceName.of(projectId, location, namespaceName))
+      .iterateAll().forEach(service -> {
+      var data = new MagpieResource.MagpieResourceBuilder(mapper, service.getName())
+        .withProjectId(projectId)
+        .withResourceType(RESOURCE_TYPE)
+        .withConfiguration(GCPUtils.asJsonNode( // Circular dependency exception while serialization of service object
+          Map.of("annotations", service.getAnnotationsMap(),
+                 "endpoints", service.getEndpointsList())))
+        .build();
+
+      emitter.emit(VersionedMagpieEnvelopeProvider.create(session, List.of(fullService() + ":service"), data.toJsonNode()));
+
+      // Discover endpoints for each service
+      discoverEndpoints(mapper, projectId, session, emitter, registrationServiceClient, location, namespaceName, service.getName());
+    });
+  }
+
+  private void discoverEndpoints(ObjectMapper mapper,
+                                 String projectId,
+                                 Session session,
+                                 Emitter emitter,
+                                 RegistrationServiceClient registrationServiceClient,
+                                 String location,
+                                 String namespaceName,
+                                 String serviceName) {
```

Comment:
service.getName() instead of 4 parameters

Reply:
agree